src/models/karaoke.py

```diff
@@ -0,0 +1,220 @@
+from dataclasses import dataclass
+from datetime import timedelta
+from functools import reduce
+from itertools import accumulate
+from typing import Sequence
+
+from src.models import SongLine
+
+from src.fonts import FontScaler
+
+import pyass
+from pyass import Style
+
+
+class KLine:
+    pass
+
+
+class KSyl:
+    pass
+
+
+@dataclass
+class KChar:
+    char: str
+    i: int
+    sylI: int
+    fadeOffset: timedelta() = timedelta()
+    karaStart: timedelta() = timedelta()
+    karaEnd: timedelta() = timedelta()
+    karaDuration: timedelta() = timedelta()
+
+    line: KLine = None
+    syl: KSyl = None
```

Comment:
Not a fan of having back references to parent objects unless its absolutely necessary...

Seems like if these variables can be removed then there's also no need for the forward declaration?

Reply:
Backreferences are more convenient when only passing in a `KChar`/`KSyl` object to functions and being able to pull parent data when making tags, like the syllable's actor, syllable's timing, line's actor transitions etc.

It could be avoided by only acting on `KSyl` and `KLine` objects and pulling data from them into a separate method instead of using backreferences, but that seems a little clunky when even the default styling needs data from the line in it too.